manifests/0000_30_cluster-api_09_admission-policies.yaml

```diff
@@ -0,0 +1,194 @@
+        - name: oldAnn
+          expression: "oldObject.metadata.?annotations.hasValue() ? oldObject.metadata.?annotations.value() : {}"
+    
+        - name: specLockedExceptAuthoritativeAPI
+          expression: >
+            object.spec.authoritativeAPI != oldObject.spec.authoritativeAPI &&
+            [
+              [object.spec.?lifecycleHooks,   oldObject.spec.?lifecycleHooks],
+              [object.spec.?metadata,         oldObject.spec.?metadata],
+              [object.spec.?providerID,       oldObject.spec.?providerID],
+              [object.spec.?providerSpec,     oldObject.spec.?providerSpec],
+              [object.spec.?taints,           oldObject.spec.?taints]
+            ].all(p,
+              p[0].hasValue()
+                ? p[1].hasValue() && p[0].value() == p[1].value()
+                : !p[1].hasValue()
+            )
+    
+      # All validations must evaluate to TRUE
+      validations:
+        # Only spec.authoritativeAPI may change
+        - expression: "object.spec == oldObject.spec || variables.specLockedExceptAuthoritativeAPI"
+          message:  "You may only modify spec.authoritativeAPI. Any other change inside .spec is not allowed."
+    
+        # Guard machine.openshift.io/* and kubernetes.io/* labels
+        - expression: >
+            !(
+              variables.newLabels.exists(k,
+                  (k.startsWith('machine.openshift.io') || k.startsWith('kubernetes.io')) &&
+                  ((k in variables.oldLabels ? variables.oldLabels[k] : '') != variables.newLabels[k])
+              ) ||
+              variables.oldLabels.exists(k,
+                  (k.startsWith('machine.openshift.io') || k.startsWith('kubernetes.io')) &&
+                  !(k in variables.newLabels)
+              )
+            )
+          message: "Cannot modify or delete any machine.openshift.io/* or kubernetes.io/* label."
+    
+        # Guard machine.openshift.io/* annotations
+        - expression: >
+            !(
+              variables.newAnn.exists(k,
+                  k.startsWith('machine.openshift.io') &&
+                  ((k in variables.oldAnn ? variables.oldAnn[k] : '') != variables.newAnn[k])
+              ) ||
+              variables.oldAnn.exists(k,
+                  k.startsWith('machine.openshift.io') &&
+                  !(k in variables.newAnn)
+              )
+            )
+          message: "Cannot modify or delete any machine.openshift.io/* annotation."
+    
+        # Param-controlled labels (labels on the CAPI machine) may change only to match the value on the CAPI Machine
+        - expression: >
+            !variables.paramLabels.exists(k,
+              ((k in variables.newLabels ? variables.newLabels[k] : '') !=
+               (k in variables.oldLabels ? variables.oldLabels[k] : ''))
+              &&
+              ((k in variables.newLabels ? variables.newLabels[k] : '') !=
+               variables.paramLabels[k])
+            )
```

Comment:
What if you went for `variables.paramLabels.all(k, variables.newLables[?k].hasValue() ? variables.newLables[k] == variables.paramLabels[k] : !variables.oldLabels[?k].hasValue())`

Check each label in the CAPI side is present and has the same value as the CAPI side, or, is missing in the old representation? Would that achieve the same thing? Would it be more succinct? 

It means you don't have to have the `orValue('')` style you have now which means you don't need to worry about whether a label is set deliberately to the empty string or not?

Reply:
Yep, I think this makes sense. 

For review context: What we're changing from is: `there exists no key in parameter labels, where that key also exists and is changing in newLabels, and that key in newLabels doesn't match the key in parameter labels`

Which... grokking `Check each label in the CAPI side is present and has the same value as the CAPI side, or, is missing in the old representation` I agree that's going to do the same thing.

Reply:
Actually... they're subtly different? (I think)

> where that key also exists and is changing in newLabels,

With my original expression, we allow mismatches e.g between `param[k]` and `newLabels[k]` when we don't change `oldLabels[k]`. 

The new one, because of the `paramLabels.all(k` would force users to either: 

Update the value of the label on the MAPI Machine to match the param (`paramLabels[k] == newLabels[k]`), or deny the policy..

Are we ok with that?

Reply:
Ahh right, yeah, so allowing it to not change makes sense. I think I was thinking that the CAPI controllers would have updated the labels in production, but this seems safe and makes sense to ratchet, LGTM